### PR TITLE
Make Dual working with Argt_with_history

### PR DESCRIPTION
**Pull request by sloriot**
Problem reported on cgal-discuss.

### CODE REVIEW

**Comment by sloriot** [comment]
@efifogel I had to use a generic file included several times because some functions are overloaded. It is somehow similar to what we did for OpenMesh. Note that the same thing can be done for the regular arrangement graph_traits (currently only 2 classes are supported).

**Comment by lrineau** [comment]
@sloriot You added the label "under discussion" to the PR. Does this mean that the PR cannot be merged as it is? I plan to test it tonight. If the test are good, shall I merge it or not? If not, please add the violet label "not yet approved". Violet labels are those blocking me from merging.

**Comment by sloriot** [comment]
I'm still waiting for @efifogel approval. However, you can still test it. I'm pretty confident the patch is good now. Tags updated.

**Comment by efifogel** [comment]
@slriot Why is this necessary? I mean, what was wrong with the previous
solution where all the 4 derived from a common template, say
Dual_arrangement_on_surface_2:

template <typename Arrangement>
class Dual_arrangement_on_surface_2 { ... };

and

1. for Arrangement_on_surface_2
template <class GeomTraits, class TopolTraits>
class Dual<Arrangement_on_surface_2<GeomTraits, TopolTraits> > :
  public Dual_arrangement_on_surface_2<Arrangement_on_surface_2<GeomTraits,
TopolTraits> >
{ ... }

2. for Arrangement_2
template <class GeomTraits, class Dcel>
class Dual<Arrangement_2<GeomTraits, TopolTraits> > :
  public Dual_arrangement_on_surface_2
    <Arrangement_2<GeomTraits,
                    Arrangement_2<GeomTraits, Dcel>::Topology_traits > >
{ ... }

3. another one for Arrangement_on_surface_with_history_2

4. and another one for Arrangement_with_history_2

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
It requires free functions for overloads for all the 4 arrangement classes (otherwise the generic free functions for `Dual<Primal>` for BGL will be called).

**Comment by efifogel** [comment]
I see, so there are 11 free functions and one class that needs to be
defined for each "version" of an arrangement.
Currently, we focus at 4 versions, but there will be more. In fact, there
are already more (i.e., Envelope_diagram_on_surface_2 in the Envelope_3
package and Arr_spherical_gaussian_map_3, which hasn't been exposed yet).
First, we need to divide graph_traits_Dual_Arrangement_2.h into 4 header
files.
Secondly, an alternative is to define 11 macros in one common file, and
then, in each of the 4 files above, have 11 calls to the various macros;
see below. I kind of prefer this variant, because it is more conventional.
What do you think?

////////
#define CGAL_DUAL_ARRANGEMENT_2_OUT_DEGREE(name)
                \
template <class T1, class
T2>                                                   \
typename boost::graph_traits<CGAL::Dual<name<T1,T2> >
>::degree_size_type       \
out_degree(typename
\
           boost::graph_traits<CGAL::Dual<name<T1,T2> >
>::vertex_descriptor v, \
           const CGAL::Dual<name<T1,T2> >&
darr)                                \
{ return darr.degree(v); }

////////
#define CGAL_DUAL_ARRANGEMENT_2_OUT_EDGES(name)
                \
template <class T1, class
T2>                                                   \
std::pair<typename
\
          boost::graph_traits<CGAL::Dual<name<T1,T2> >
>::out_edge_iterator,    \

typename                                                              \
          boost::graph_traits<CGAL::Dual<name<T1,T2> >
>::out_edge_iterator>    \
out_edges(typename
\
          boost::graph_traits<CGAL::Dual<name<T1,T2> > >::vertex_descriptor
v,  \
          const CGAL::Dual<name<T1,T2> >&
darr)                                 \
{ return std::make_pair(darr.out_edges_begin(v), darr.out_edges_end(v)); }

Then call:

CGAL_DUAL_ARRANGEMENT_2_OUT_DEFREE(Arrangement_on_surface_2)
CGAL_DUAL_ARRANGEMENT_2_OUT_EDGES(Arrangement_on_surface_2)
...

CGAL_DUAL_ARRANGEMENT_2_OUT_DEGREE(Arrangement_2)
CGAL_DUAL_ARRANGEMENT_2_OUT_EDGES(Arrangement_2)
...

CGAL_DUAL_ARRANGEMENT_2_OUT_DEGREE(Arrangement_on_surface_with_history_2)
CGAL_DUAL_ARRANGEMENT_2_OUT_EDGES(Arrangement_on_surface_with_history_2)
...

CGAL_DUAL_ARRANGEMENT_2_OUT_DEGREE(Arrangement_with_history_2)
CGAL_DUAL_ARRANGEMENT_2_OUT_EDGES(Arrangement_with_history_2)
...

   1. out_degree
   2. out_edges
   3. source
   4. target
   5. in_degree
   6. in_edges
   7. degree
   8. num_vertices
   9. vertices
   10. num_edges
   11. edges

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
The pb with macros is that as soon as there is a compilation error, it is getting hard to understand what's going on. With the header file, it simply points to the problematic line. In addition the file is much more readable.

**Comment by efifogel** [comment]
1. The macros are also more flexible. Imagine a "version" of an arrangement
that requires an implementation that is slightly different than the one
used by all other. Then, you simply introduce the special definition.

2. Also, we can have one additional macro:

#define CGAL_DUAL_ARRANGEMENT_2_FNC_DEFS(name) \
CGAL_DUAL_ARRANGEMENT_2_OUT_DEFREE(name)       \
CGAL_DUAL_ARRANGEMENT_2_OUT_EDGES(name)        \
...

It reduces the code back to a one-liner.

3. The compilation errors are hard to decipher anyway, and the
unconventional coding makes it, actually, harder, to handle generally
speaking.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
You are the maintainer of the package so it's up to you. I can close this PR or you can push directly the changes in this one.

**Comment by efifogel** [comment]
ok, let me role it my head for a couple of days and then I'll act.
Let me know in case you prefer that I'll create and edit a branch of mine
instead of yours. (I'm going to split the header anyway.)
Thanks for the effort!!!
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
Do what's the best for you.

**Comment by efifogel** [comment]
ok, I've made the changes and pushed them back to your branch.
1. I somehow broke backward compatibility. I've introduced 4 header files:

   1. graph_traits_dual_arrangement_2.h
   2. graph_traits_dual_arrangement_on_surface_2.h
   3. graph_traits_dual_arrangement_on_surface_with_history_2.h
   4. graph_traits_dual_arrangement_with_history_2.h

The first (kind of) replaces the former graph_traits_Dual_Arrangement_2.h.
As you can see, the name of the old and obsolete one and the new differ in
letter case only.

2. I still need to report the changes. How do I do that now days? (Couldn't
find changes.html file...)
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by afabri** [comment]
For Polyhedron and Surface_mesh, we include the graph_traits file for class X directly from X.h

**Comment by efifogel** [comment]
Why? It seems like an awful thing to do.
1. It unnecessarily increases the compilation units for those apps that
include X.h and do not need the the graph_traits..
2. But even worse, it may cause cpp files to be recompiled even if they
don't really need to (when the graph traits is touched) and compilation is
a major time consumer in our case.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
There are errors in Travis:
https://travis-ci.org/CGAL/cgal/jobs/423411098#L1862
```
[ 33%] Building CXX object CMakeFiles/envelope_triangles.dir/envelope_triangles.cpp.o
In file included from /home/travis/build/CGAL/cgal/Envelope_3/examples/Envelope_3/envelope_planes.cpp:7:
In file included from /home/travis/build/CGAL/cgal/Envelope_3/examples/Envelope_3/../../include/CGAL/envelope_3.h:30:
/home/travis/build/CGAL/cgal/Envelope_3/examples/Envelope_3/../../include/CGAL/Envelope_3/Envelope_divide_and_conquer_3.h:49:10: fatal error: 
      'CGAL/graph_traits_Dual_Arrangement_2.h' file not found
#include <CGAL/graph_traits_Dual_Arrangement_2.h>
         ^
In file included from /home/travis/build/CGAL/cgal/Envelope_3/examples/Envelope_3/envelope_triangles.cpp:8:
In file included from /home/travis/build/CGAL/cgal/Envelope_3/examples/Envelope_3/../../include/CGAL/envelope_3.h:30:
/home/travis/build/CGAL/cgal/Envelope_3/examples/Envelope_3/../../include/CGAL/Envelope_3/Envelope_divide_and_conquer_3.h:49:10: fatal error: 
      'CGAL/graph_traits_Dual_Arrangement_2.h' file not found
#include <CGAL/graph_traits_Dual_Arrangement_2.h>
         ^
1 error generated.
```

**Comment by efifogel** [comment]
This is fixed, committed, and pushed now.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

File: Installation/CHANGES.md [hunk 1]
```diff
@@ -252,6 +252,23 @@ g
...
+    `Arrangement_on_surface_with_history_2`, and
+    `Arrangement_with_history_2`.
+
+-   **Breaking change**: The old `<CGAL/graph_traits_Dual_Arrangement_2.h>`
+    header file has been replaced by the four header files below; each defines
+    the `graph_traits` for dual of the corresponding arrangement type.
+    `<CGAL/graph_traits_dual_arrangement_on_surface_2.h>`,
+    `<CGAL/graph_traits_dual_arrangement_2.h>`,
+    `<CGAL/graph_traits_dual_arrangement_on_surface_with_history_2.h>`, and
+    `<CGAL/graph_traits_dual_arrangement_with_history_2.h`.
```
[the remark this answers is not in the document]
Comment (lrineau) [reply Installation/CHANGES.md]: Actually, I was wrong, and a15560a7b1929929219405fd6385a032471236af did add the changes for this PR.

**Pull request merged**